// shared/dynamodb.js
import { createAwsSigner } from 'sign-aws-requests'

const makeRequest = async ({ region, secret, id }, params, type) => {
	const sign = createAwsSigner({
		config: {
			service: 'dynamodb',
			region: region,
			secretAccessKey: secret,
			accessKeyId: id
		}
	})
	const request = {
		url: `https://dynamodb.${region}.amazonaws.com`,
		method: 'POST',
		headers: {
			'Content-Type': 'application/x-amz-json-1.0',
			'X-Amz-Target': `DynamoDB_20120810.${type}`,
			Host: `dynamodb.${region}.amazonaws.com`
		},
		body: JSON.stringify(params)
	}
	const { authorization } = await sign(request)
	request.headers.Authorization = authorization
	const response = await fetch(request.url, {
		method: request.method,
		headers: request.headers,
		body: request.body
	})
	return response.json()
}

export function dynamodb(credentials) {
	return {
		getItem: async params => makeRequest(credentials, params, 'GetItem'),
		putItem: async params => makeRequest(credentials, params, 'PutItem'),
		query: async params => makeRequest(credentials, params, 'Query')
	}
}
