// @braidedjs/worker/dynamodb.js
// TODO: turn this into `@saibotsivad/dynamodb` or similar
import { createAwsSigner } from 'sign-aws-requests'

const makeRequest = async ({ region, secretAccessKey, accessKeyId }, { httpFetch }, params, type) => {
	const sign = createAwsSigner({
		config: {
			service: 'dynamodb',
			region: region,
			secretAccessKey,
			accessKeyId
		}
	})
	const request = {
		url: `https://dynamodb.${region}.amazonaws.com`,
		method: 'POST',
		headers: {
			'Content-Type': 'application/x-amz-json-1.0',
			'X-Amz-Target': `DynamoDB_20120810.${type}`,
			Host: `dynamodb.${region}.amazonaws.com`
		},
		body: JSON.stringify(params)
	}
	const { authorization } = await sign(request)
	request.headers.Authorization = authorization
	console.log(JSON.stringify({ message: 'http request', request }))
	const response = await httpFetch(request.url, {
		method: request.method,
		headers: request.headers,
		body: request.body
	})
	return response.json()
}

export function dynamodb(credentials, configuration = { httpFetch: fetch }) {
	return {
		batchGetItem: async params => makeRequest(credentials, configuration, params, 'BatchGetItem'),
		batchWriteItem: async params => makeRequest(credentials, configuration, params, 'BatchWriteItem'),
		deleteItem: async params => makeRequest(credentials, configuration, params, 'DeleteItem'),
		getItem: async params => makeRequest(credentials, configuration, params, 'GetItem'),
		putItem: async params => makeRequest(credentials, configuration, params, 'PutItem'),
		query: async params => makeRequest(credentials, configuration, params, 'Query')
	}
}
